Userland/Libraries/LibCore/Stream.h

```diff
@@ -0,0 +1,846 @@
+    KResult set_blocking(bool enabled) override { return m_helper.set_blocking(enabled); }
+    KResult set_close_on_exec(bool enabled) override { return m_helper.set_close_on_exec(enabled); }
+
+    virtual ~UDPSocket() override { close(); }
+
+private:
+    UDPSocket() { }
+
+    void setup_notifier()
+    {
+        VERIFY(is_open());
+
+        m_helper.setup_notifier();
+        m_helper.notifier()->on_ready_to_read = [this] {
+            if (on_ready_to_read)
+                on_ready_to_read();
+        };
+    }
+
+    PosixSocketHelper m_helper { Badge<UDPSocket> {} };
+};
+
+class LocalSocket final : public Socket {
+public:
+    static KResultOr<LocalSocket> connect(String const& path);
+
+    LocalSocket(LocalSocket&& other)
+        : Socket(static_cast<Socket&&>(other))
+        , m_helper(move(other.m_helper))
+    {
+        if (is_open())
+            setup_notifier();
+    }
+
+    LocalSocket& operator=(LocalSocket&& other)
+    {
+        Socket::operator=(static_cast<Socket&&>(other));
+        m_helper = move(other.m_helper);
+        if (is_open())
+            setup_notifier();
+
+        return *this;
+    }
+
+    virtual bool is_readable() const override { return is_open(); }
+    virtual bool is_writable() const override { return is_open(); }
+    virtual KResultOr<size_t> read(Bytes buffer) override { return m_helper.read(buffer); }
+    virtual KResultOr<size_t> write(ReadonlyBytes buffer) override { return m_helper.write(buffer); }
+    virtual bool is_eof() const override { return m_helper.is_eof(); }
+    virtual bool is_open() const override { return m_helper.is_open(); }
+    virtual void close() override { m_helper.close(); }
+    virtual KResultOr<size_t> pending_bytes() const override { return m_helper.pending_bytes(); }
+    virtual KResultOr<bool> can_read_without_blocking(int timeout = 0) const override { return m_helper.can_read_without_blocking(timeout); }
+    KResult set_blocking(bool enabled) override { return m_helper.set_blocking(enabled); }
+    KResult set_close_on_exec(bool enabled) override { return m_helper.set_close_on_exec(enabled); }
+
+    virtual ~LocalSocket() { close(); }
+
+private:
+    LocalSocket() { }
+
+    void setup_notifier()
+    {
+        VERIFY(is_open());
+
+        m_helper.setup_notifier();
+        m_helper.notifier()->on_ready_to_read = [this] {
+            if (on_ready_to_read)
+                on_ready_to_read();
+        };
+    }
+
+    PosixSocketHelper m_helper { Badge<LocalSocket> {} };
+};
+
+// Buffered stream wrappers
+
+template<typename T>
+concept IsStream = IsBaseOf<Stream, T>;
+template<typename T>
+concept IsSeekableStream = IsBaseOf<SeekableStream, T>;
+template<typename T>
+concept IsSocket = IsBaseOf<Socket, T>;
+
+template<typename T>
+class BufferedHelper {
+    AK_MAKE_NONCOPYABLE(BufferedHelper);
+
+public:
+    template<typename U>
+    BufferedHelper(Badge<U>, T stream, ByteBuffer buffer) requires(IsBaseOf<Stream, U>)
```

Comment:
We usually take the badge as the last parameter.

Reply:
Really? I've seen many examples where we take the badge first, or is this a class-specific thing?

Reply:
Huh.
I guess I was misinformed about it, convention seems to be to take the badge _first_.
your code is fine then